AMP-59355 - add kotlin advanced topic (#315)
Amplitude-Developer-Docs/amplitude-dev-center: docs/data/sdks/android-kotlin/index.md

    new SegmentDestinationPlugin(this, SEGMENT_WRITE_KEY)
)
```

## Advanced Topic

### User sessions

A session on Android is a period of time that a user has the app in the foreground.

Amplitude groups events together by session. Events that are logged within the same session have the same `session_id`. Sessions are handled automatically so you don't have to manually call `startSession()` or `endSession()`.

You can adjust the time window for which sessions are extended.

=== "Java"

    ```
    amplitude = AmplitudeKt.Amplitude(AMPLITUDE_API_KEY, getApplicationContext(), configuration -> {
        configuration.setMinTimeBetweenSessionsMillis(1000);
        return Unit.INSTANCE;
    });
    ```
=== "Kotlin"

    ```kotlin
    amplitude = Amplitude(
            Configuration(
                apiKey = AMPLITUDE_API_KEY,
                context = applicationContext,
                minTimeBetweenSessionsMillis = 10000
            )

>         )

Reviewer: Minor, style, align to left?

```
amplitude = Amplitude(
    Configuration(
        apiKey = AMPLITUDE_API_KEY,
        context = applicationContext,
        minTimeBetweenSessionsMillis = 10000
    )
)
```

    ```

> By default, Amplitude will auto send the '[Amplitude] Start Session' and '[Amplitude] End Session'. Even though these events aren't sent, sessions are still tracked by using `session_id`.

Reviewer 2: This seems contradictory? I also think it might be missing a word, maybe _events?_ -- "the '[Amplitude] Start Session' and '[Amplitude] End Session'. _events_"

Author: More explanation:
Logging the session events are default behavior, but customers can disable this. So if the customer disables the trackingSessionEvents, they will not receive `'[Amplitude] Start Session' and '[Amplitude] End Session'` events. 
But each event will still have a session_id value unless there has a setting to force the session_id = -1.

Author: Let me know if you have better words to clarify those all.

You can also disable those session events.

=== "Java"

    ```
    amplitude = AmplitudeKt.Amplitude(AMPLITUDE_API_KEY, getApplicationContext(), configuration -> {
        configuration.setTrackingSessionEvents(false);
        return Unit.INSTANCE;
    });
    ```
=== "Kotlin"

    ```kotlin
    amplitude = Amplitude(
            Configuration(
                apiKey = AMPLITUDE_API_KEY,
                context = applicationContext,
                trackingSessionEvents = false
            )
        )
    ```

> You can use the helper method getSessionId to get the value of the current sessionId.

Reviewer: Minor, `getSessionId`

> You can use the helper method getSessionId to get the value of the current sessionId.

Reviewer 2: ```suggestion
You can use the helper method `getSessionId` to get the value of the current `sessionId`.
```

=== "Java"

    ```java
    long sessionId = amplitude.getSessionId();
    ```

=== "Kotlin"

    ```kotlin
    val sessionId = amplitude.sessionId;
    ```

!!!note

>     For Android API level 14 and above, a new session is created when the app comes back into the foreground after being in the background for five or more minutes or when the last event was logged

Reviewer 2: ```suggestion
    For Android API level 14 and higher, a new session is created when the app comes back into the foreground after being in the background for more than five minutes or when the last event was logged.
```

Clarifying question -- so a new session is created after 5 minutes of the app being backgrounded OR after 5 minutes pass since the last logged event?

Author: I just noticed we have different behavior in the new android SDK. I will change some words here. 
It should be pass x minutes since the last logged event.

     (whichever occurred last).
     Otherwise, the background event logged is part of the current session.

>       Note that you can define your own session expiration time by calling setMinTimeBetweenSessionsMillis(timeout), where the timeout input is in milliseconds.

Reviewer: Not sure whether we want to use inline code for function name in notes, like `setMinTimeBetweenSessionsMillis(timeout)`, leave it to you to decide.

>       Note that you can define your own session expiration time by calling setMinTimeBetweenSessionsMillis(timeout), where the timeout input is in milliseconds.

Reviewer 2: ```suggestion
      You can define your own session expiration time by calling `setMinTimeBetweenSessionsMillis(timeout)`, where the timeout input is in milliseconds.
```
We do put function names in backticks.

>     For Android API level 13 and below, foreground tracking is not available so a new session is automatically started when an event is logged 30 minutes or more after the last logged event. If another event is logged within 30 minutes, it will extend the current session. Note that you can define your own session expiration time here as well by calling setSessionTimeoutMillis(timeout), where the timeout input is in milliseconds. Also note that enableForegroundTracking(getApplication) is still safe to call for Android API level 13 and below even though it's not available.

Reviewer 2: ```suggestion
    For Android API level 13 and lower, foreground tracking is not available so a new session is automatically started when an event is logged 30 minutes or more after the last logged event. If another event is logged within 30 minutes, it extends the current session. You can define your own session expiration time by calling `setSessionTimeoutMillis(timeout)`, where the timeout input is in milliseconds. It's safe to call `enableForegroundTracking(getApplication)` for Android API level 13 and lower even though it's not available.
```

Is "Android API level" the correct term, or should we be using "version"?

Author: Yes, it should be "Android API Level".

    amplitude.reset()
    ```

### Disable tracking

By default the Android SDK tracks several user properties such as `carrier`, `city`, `country`, `ip_address`, `language`, and `platform`.
Use the provided `TrackingOptions` interface to customize and toggle individual fields.

To use the `TrackingOptions` interface, import the class.

=== "Java"

    ```java
    import com.amplitude.android.TrackingOptions
    ```

=== "Kotlin"

    ```kotlin
    import com.amplitude.android.TrackingOptions
    ```

Before initializing the SDK with your apiKey, create a `TrackingOptions` instance with your configuration and set it on the SDK instance.

=== "Java"

    ```java
        TrackingOptions trackingOptions = new TrackingOptions();
        trackingOptions.disableCity().disableIpAddress().disableLatLng();

        // init instance
        amplitude = AmplitudeKt.Amplitude(AMPLITUDE_API_KEY, getApplicationContext(), configuration -> {
            configuration.setTrackingOptions(trackingOptions);
            return Unit.INSTANCE;
        });
    ```

=== "Kotlin"

    ```kotlin
    val trackingOptions = TrackingOptions()
    trackingOptions.disableCity().disableIpAddress().disableLatLng()

>         amplitude = Amplitude(

Reviewer: Minor, indentation alignment

            Configuration(
                apiKey = AMPLITUDE_API_KEY,
                context = applicationContext,
                trackingOptions = trackingOptions
        )
    )
    ```

> Tracking for each field can be individually controlled, and has a corresponding method (for example, `disableCountry`, `disableLanguage`).

Reviewer 2: ```suggestion
You can control tracking for each field individually with the field's corresponding method (for example, `disableCountry`, `disableLanguage`).
```
Please confirm that I clarified without introducing inaccuracy.

| <div class="big-column">Method</div> | Description |
| --- | --- |
| `disableAdid()` | Disable tracking of Google ADID |
| `disableCarrier()` | Disable tracking of device's carrier |
| `disableCity()` | Disable tracking of user's city |
| `disableCountry()` | Disable tracking of user's country |
| `disableDeviceBrand()` | Disable tracking of device brand |
| `disableDeviceModel()` | Disable tracking of device model |
| `disableDma()` | Disable tracking of user's designated market area (DMA). |
| `disableIpAddress()` | Disable tracking of user's IP address |
| `disableLanguage()` | Disable tracking of device's language |
| `disableLatLng()` | Disable tracking of user's current latitude and longitude coordinates |
| `disableOsName()` | Disable tracking of device's OS Name |
| `disableOsVersion()` | Disable tracking of device's OS Version |
| `disablePlatform()` | Disable tracking of device's platform |
| `disableRegion()` | Disable tracking of user's region. |
| `disableVersionName()` | Disable tracking of your app's version name |

!!!note

>     The `TrackingOptions` only prevents default properties from being tracked on newly created projects, where data has not yet been sent. If you have a project with existing data that you would like to stop collecting the default properties for, please get help in the [Amplitude Community](https://community.amplitude.com/). Note that the existing data **is not** deleted.

Reviewer 2: ```suggestion
    Using `TrackingOptions` only prevents default properties from being tracked on newly created projects, where data has not yet been sent. If you have a project with existing data that you want to stop collecting the default properties for, get help in the [Amplitude Community](https://community.amplitude.com/). Disabling tracking doesn't delete any existing data in your project.
```

        configuration.setEnableCoppaControl(true); //Disables ADID, city, IP, and location tracking
        return Unit.INSTANCE;
    });
    client.enableCoppaControl();
    ```

=== "Kotlin"

    ```kotlin
    amplitude = Amplitude(
        Configuration(
            apiKey = AMPLITUDE_API_KEY,
            context = applicationContext,
            enableCoppaControl = true //Disables ADID, city, IP, and location tracking
        )
    )
    ```

### Advertiser ID

Advertiser ID (also referred to as IDFA) is a unique identifier provided by the iOS and Google Play stores. As it's unique to every person and not just their devices, it's useful for mobile attribution.
 [Mobile attribution](https://www.adjust.com/blog/mobile-ad-attribution-introduction-for-beginners/) is the attribution of an installation of a mobile app to its original source (such as ad campaign, app store search).
 Mobile apps need permission to ask for IDFA, and apps targeted to children can't track at all. Consider IDFV or device ID when IDFA isn't available.

Follow these steps to use Android Ad ID.

!!!warning "Google Ad ID and Tracking Warning"

>     As of April 1, 2022, Google allows users to opt out of Ad ID tracking. Ad ID may return null or error. You can use am alternative ID called [App Set ID](#app-set-id), which is unique to every app install on a device. [Learn more](https://support.google.com/googleplay/android-developer/answer/6048248?hl=en)

Reviewer 2: Indent: this whole text block must be indented 4 spaces for the warning block to work. Double-check in your editor, please.

Author: It looks right in my editor and on the web. 
I think it's the github issue.
<img width="1196" alt="Screen Shot 2022-10-07 at 10 00 20 AM" src="https://user-images.githubusercontent.com/13028329/194611318-bdbd9280-5cfb-4e79-932d-aa3c091f13a5.png">
<img width="770" alt="Screen Shot 2022-10-07 at 10 17 15 AM" src="https://user-images.githubusercontent.com/13028329/194613951-e94e97e1-f855-4fa1-bae3-9712d56d130b.png">

Reviewer 2: Thank you so much for confirming, sorry for the trouble!

        configuration.setUseAdvertisingIdForDeviceId(true);
        return Unit.INSTANCE;
    });
    ```

=== "Kotlin"

    ```kotlin
    amplitude = Amplitude(
        Configuration(
            apiKey = AMPLITUDE_API_KEY,
            context = applicationContext,
            useAdvertisingIdForDeviceId = true
        )
    )
    ```

### App set ID

App set ID is a unique identifier for each app install on a device. App set ID is reset by the user manually when they uninstall the app, or after 13 months of not opening the app.
 Google designed this as a privacy-friendly alternative to Ad ID for users who want to opt out of stronger analytics.

 To use App Set ID, follow these steps.

1. Add `play-services-appset` as a dependency. For versions earlier than 2.35.3, use `'com.google.android.gms:play-services-appset:16.0.0-alpha1'`

    ```bash
    dependencies {
        implementation 'com.google.android.gms:play-services-appset:16.0.2'
    }
    ```

2. Enable to use app set ID as Device ID.

=== "Java"

    ```java
    amplitude = AmplitudeKt.Amplitude(BuildConfig.AMPLITUDE_API_KEY, getApplicationContext(), configuration -> {

>         configuration.setuseAppSetIdForDeviceId(true);

Reviewer: Minor, `setuseAppSetIdForDeviceId` -> `setUseAppSetIdForDeviceId`
